hibernate-core/src/main/java/org/hibernate/engine/spi/ActionQueue.java

```diff
@@ -1170,82 +1151,58 @@ public void sort(List<AbstractEntityInsertAction> insertions) {
 					BatchIdentifier nextBatchIdentifier = latestBatches.get( j );
 
-					if ( nextBatchIdentifier.hasAnyParentEntityNames( batchIdentifier ) ) {
-						nextBatchIdentifier.parent = batchIdentifier;
-					}
-					if ( batchIdentifier.hasAnyChildEntityNames( nextBatchIdentifier ) ) {
-						nextBatchIdentifier.parent = batchIdentifier;
+					if ( nextBatchIdentifier.hasAnyParentEntityNames( batchIdentifier ) || batchIdentifier.hasAnyChildEntityNames( nextBatchIdentifier ) ) {
+						nextBatchIdentifier.getParents().add( batchIdentifier );
+						batchIdentifier.getChildren().add( nextBatchIdentifier );
+						batchesWithoutParents.remove( nextBatchIdentifier );
 					}
 				}
 			}
 
-			boolean sorted = false;
-
-			long maxIterations = latestBatches.size() * latestBatches.size();
-			long iterations = 0;
-
-			sort:
-			do {
-				// Examine each entry in the batch list, sorting them based on parent/child association
-				// as depicted by the dependency graph.
-				iterations++;
-
-				for ( int i = 0; i < latestBatches.size(); i++ ) {
-					BatchIdentifier batchIdentifier = latestBatches.get( i );
+			final List<BatchIdentifier> sortedBatches = new ArrayList<>( latestBatches.size() );
 
-					// Iterate next batches and make sure that children types are after parents.
-					// Since the outer loop looks at each batch entry individually and the prior loop will reorder
-					// entries as well, we need to look and verify if the current batch is a child of the next
-					// batch or if the current batch is seen as a parent or child of the next batch.
-					for ( int j = i + 1; j < latestBatches.size(); j++ ) {
-						BatchIdentifier nextBatchIdentifier = latestBatches.get( j );
+			LinkedList<BatchIdentifier> queue = new LinkedList<>(batchesWithoutParents);
 
-						if ( batchIdentifier.hasParent( nextBatchIdentifier ) ) {
-							if ( nextBatchIdentifier.hasParent( batchIdentifier ) ) {
-								//cycle detected, no need to continue
-								break sort;
-							}
-
-							latestBatches.remove( batchIdentifier );
-							latestBatches.add( j, batchIdentifier );
-
-							continue sort;
-						}
+			// Topological Sorting (i.e. Kahn algorithm)
+			while ( ! queue.isEmpty() ) {
+				BatchIdentifier batch = queue.poll();
+				sortedBatches.add( batch );
+				for ( BatchIdentifier child : batch.getChildren() ) {
+					child.getParents().remove( batch );
+					if ( child.getParents().isEmpty() ) {
+						queue.push( child );
 					}
 				}
-				sorted = true;
 			}
-			while ( !sorted && iterations <= maxIterations );
 
-			if ( iterations > maxIterations ) {
-				LOG.warn( "The batch containing " + latestBatches.size() + " statements could not be sorted after " + maxIterations + " iterations. " +
-								"This might indicate a circular entity relationship." );
+			if ( sortedBatches.size() < latestBatches.size() ) {
+				LOG.warn( "The batch containing " + latestBatches.size() + " statements could not be sorted because of circular dependencies");
```

Comment:
we can easily get the cycle participants by computing the two collection diff, if needed.